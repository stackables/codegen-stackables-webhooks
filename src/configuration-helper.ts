interface ConfigurationOptions {
    generatedFile: string
    accountSlug: string
    introspectionToken: string
}

export function getConfiguration(opts: ConfigurationOptions) {
    const schemaUrl = `https://data.stackables.io/${opts.accountSlug}?introspection=${opts.introspectionToken}`
    return {
        schema: schemaUrl,
        extensions: {
            codegen: {
                generates: {
                    [opts.generatedFile]: {
                        documents: [
                            {
                                [schemaUrl]: {
                                    loader: 'codegen-stackables-webhooks',
                                },
                            },
                        ],
                        plugins: ['typescript', 'typescript-operations', 'stackables-webhooks'],
                        config: {
                            preResolveTypes: true,
                            onlyOperationTypes: true,
                            fragmentRegistryName: 'StackablesEvents',
                            fragmentRegistryDirective: 'register'
                        }
                    }
                }
            }
        }
    }
}
