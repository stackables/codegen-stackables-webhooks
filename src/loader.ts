import { parse } from "graphql";
import { gql, GraphQLClient } from "graphql-request";

export default async function (
	endpoint: string,
	_: unknown,
	conf: Record<string, { bearerToken: string }>
) {
	const graphQLClient = new GraphQLClient(endpoint, {
		headers: {
			authorization: `Bearer ${conf[endpoint].bearerToken}`,
		},
	});

	const query = gql`
		query CodeGenWebhooks {
			Webhook: _Webhook {
				list {
					name
					fragment
					entity {
						id
					}
				}
			}
		}
	`;

	const data = await graphQLClient.request(query);
	const fragments = data.Webhook.list;

	return parse(
		fragments
			.map(
				(f: any) =>
					`fragment ${f.name} on ${f.entity.id} @register ${f.fragment}`
			)
			.join("\n")
	);
}
