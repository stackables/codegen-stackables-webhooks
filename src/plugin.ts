import { PluginFunction, Types } from '@graphql-codegen/plugin-helpers';
import { capitalCase } from 'change-case-all';
import { concatAST, DirectiveNode, DocumentNode, GraphQLSchema, visit } from 'graphql';

interface NamedOperationsObjectPluginConfig {
  fragmentRegistryName?: string
  fragmentRegistryDirective?: string
  omitOperationSuffix?: boolean
}

function filterByDirective(directives: readonly DirectiveNode[] | undefined, needed: string | undefined) {
  // if no restriction specified then list all
  if (!needed) return true

  // TODO: Investigate under what circumstances can it actually be undefined?
  /* istanbul ignore next */
  if (!directives) return false

  for (let index = 0; index < directives.length; index++) {
    const directive = directives[index]

    if (directive.name.value === needed) {
      return true
    }
  }

  // if nothing matched then we ignore this fragment
  return false
}

export const plugin: PluginFunction<NamedOperationsObjectPluginConfig, string> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: NamedOperationsObjectPluginConfig
) => {

  const fragmentRegistryName = config.fragmentRegistryName || 'FragmentRegistry'

  const docs = documents.map(v => v.document) as DocumentNode[]
  const allAst = concatAST(docs);
  const fragments: string[] = []

  visit(allAst, {
    FragmentDefinition: node => {
      if (filterByDirective(node.directives, config.fragmentRegistryDirective)) {
        fragments.push(node.name.value)
      }
    },
  });

  const suffix = (config.omitOperationSuffix) ? '' : 'Fragment'

  const source = []

  // Wrapper type
  source.push(`export type StackablesEnvelope<T> = { 
    version: "0.1"
    payload: T
    params: any
    callback: {
      url: string
      token: string
    }
  }`)

  // Handler functions
  fragments.map(fragment => {
    source.push(`export type ${fragment}Handler = StackablesEnvelope<${fragment}${suffix}>`)
  })

  source.push(`export type ${fragmentRegistryName} = {
${fragments.map(fragment => `  'stackables.webhook.${fragment}': ${fragment}Handler`).join('\n')}
  }`)

  return source.join('\n')
};
