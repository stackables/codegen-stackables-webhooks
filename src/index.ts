export { getConfiguration } from './configuration-helper'
export { plugin } from './plugin'
import loader from './loader'

export default loader