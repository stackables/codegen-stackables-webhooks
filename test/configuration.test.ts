import { getConfiguration } from '../src'
import { buildClientSchema, getIntrospectionQuery } from 'graphql';
import axios from 'axios'

test('Configuration snapshot', () => {
    const config = getConfiguration({
        accountSlug: 'stackables-demo:prod/example1',
        generatedFile: './src/generated/webhooks.ts',
        introspectionToken: 'a89751eb-ef0d-4663-9530-fa683b9d8483',
    })

    expect(config).toMatchSnapshot()
})

test('Configuration introspection is working correctly', async () => {
    const { schema } = getConfiguration({
        accountSlug: 'stackables-demo:prod/example1',
        generatedFile: './src/generated/webhooks.ts',
        introspectionToken: 'a89751eb-ef0d-4663-9530-fa683b9d8483',
    })

    const { data } = await axios({
        url: schema,
        method: 'post',
        data: {
            query: getIntrospectionQuery()
        }
    })

    buildClientSchema(data.data)
})