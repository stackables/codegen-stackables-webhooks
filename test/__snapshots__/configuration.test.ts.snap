// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configuration snapshot 1`] = `
Object {
  "extensions": Object {
    "codegen": Object {
      "generates": Object {
        "./src/generated/webhooks.ts": Object {
          "config": Object {
            "fragmentRegistryDirective": "register",
            "fragmentRegistryName": "StackablesEvents",
            "onlyOperationTypes": true,
            "preResolveTypes": true,
          },
          "documents": Array [
            Object {
              "https://data.stackables.io/stackables-test/webhooks?introspection=a89751eb-ef0d-4663-9530-fa683b9d8483": Object {
                "loader": "stackables-webhooks/loader",
              },
            },
          ],
          "plugins": Array [
            "typescript",
            "typescript-operations",
            "stackables-webhooks",
          ],
        },
      },
    },
  },
  "schema": "https://data.stackables.io/stackables-test/webhooks?introspection=a89751eb-ef0d-4663-9530-fa683b9d8483",
}
`;
